snap/snapcraft.yaml

```diff
@@ -0,0 +1,204 @@
+name: docker
+version: 17.03.1-ce-0
+summary: Docker Linux container runtime
+description: Docker complements kernel namespacing with a high-level API which operates at the process level. It runs unix processes with strong guarantees of isolation and repeatability across servers.
+
+confinement: strict
+grade: stable
+
+plugs:
+  support:
+    interface: docker-support
+  privileged:
+    interface: docker-support
+    privileged-containers: true
+  docker-cli:
+    interface: docker
+slots:
+  docker-daemon:
+    interface: docker
+
+apps:
+  docker:
+    command: docker-wrapper docker
+    plugs:
+      - docker-cli
+      - network
+      - home
+
+  dockerd:
+    command: dockerd-wrapper
+    daemon: simple
+    plugs:
+      - network-bind
+      - firewall-control
+      - support
+      - privileged
+    slots:
+      - docker-daemon
+
+  # TODO determine if we can "provide" a "docker-compose" in PATH directly (since this'll show up as "docker.compose" in the user PATH)
+  compose:
+    command: docker-wrapper docker-compose
+    plugs:
+      - docker-cli
+      - network
+      - home
+  help:
+    command: help
+
+parts:
+  snappy-bins:
+    plugin: shell
+    shell: bash
+    shell-flags: ['-ex']
+    shell-command: |
+      install -d "$SNAPCRAFT_PART_INSTALL/bin"
+      install -t "$SNAPCRAFT_PART_INSTALL/bin" "$SNAPDIR"/bin/*
+    stage-packages:
+      - mount
+
+  go:
+    plugin: gobuild
+    source: https://storage.googleapis.com/golang/go1.7.5.src.tar.gz
+    source-type: tar
+    snap:
+      - -*
+    build-packages:
+      - golang-go
+      - rsync
+      - build-essential
+      - netbase
+      - python3-pyelftools
+
+  docker:
+    plugin: shell
+    source: https://github.com/docker/docker/archive/v17.03.1-ce.tar.gz
+    source-type: tar
+    shell: bash
+    shell-flags: ['-eux', '-o', 'pipefail']
+    shell-command: |
+      export GOROOT="$SNAPDIR/parts/go/install"
+      export PATH="$GOROOT/bin:$PATH"
+      go version
+
+      source "$SNAPDIR/prep-docker-build.sh"
+      
+      ./hack/make.sh dynbinary
+      
+      ls -l bundles/latest/*/
+      
+      clientBin='bundles/latest/dynbinary-client/docker'
+      daemonBin='bundles/latest/dynbinary-daemon/dockerd'
+      "$clientBin" -v
+      "$daemonBin" -v
+      "$clientBin" -v | grep -q -- "$DOCKER_GITCOMMIT"
+      "$daemonBin" -v | grep -q -- "$DOCKER_GITCOMMIT"
+      
+      install -d "$SNAPCRAFT_PART_INSTALL/bin"
+      install -T "$(readlink -f "$clientBin")" "$SNAPCRAFT_PART_INSTALL/bin/docker"
+      install -T "$(readlink -f "$daemonBin")" "$SNAPCRAFT_PART_INSTALL/bin/dockerd"
+    after:
+      - go
+    build-packages:
+      - btrfs-tools
+      - gcc
+      - libc6-dev
+      - libdevmapper-dev
+      - libltdl-dev
+      - patch
+      - git
+      - pkg-config
+    stage-packages:
+      - aufs-tools
+
+  containerd:
+    plugin: shell
+    source: https://github.com/docker/containerd/archive/4ab9917febca54791c5f071a9d1f404867857fcc.tar.gz
```

Comment:
(And `tini` _must_ be static, since it gets bind-mounted into `/dev/init` inside containers which likely won't have the same libc as our snap environment)